src/gui/qgsfilewidget.h

```diff
@@ -116,6 +125,86 @@ class GUI_EXPORT QgsFileWidget : public QWidget
      */
     void setReadOnly( bool readOnly );
 
+    /**
+     * Set \a storageType storage type unique identifier as defined in QgsExternalStorageRegistry or
+     * null QString if there is no storage defined.
+     * If no external storage has been defined, QgsFileWidget will only update file path according to
+     * selected files.
+     * \see storageType
+     * \since QGIS 3.22
+     */
+    void setStorageType( const QString &storageType );
+
+    /**
+     * Returns storage type unique identifier as defined in QgsExternalStorageRegistry.
+     * Returns null QString if there is no storage defined, only file selection.
+     * \see setStorageType
+     * \since QGIS 3.22
+     */
+    QString storageType() const;
+
+    /**
+     * Returns external storage used to store selected file names, nullptr if none have been defined.
+     * If no external storage has been defined, QgsFileWidget will only update file path according to
+     * selected files.
+     * \see setStorageType
+     * \since QGIS 3.22
+     */
+    QgsExternalStorage *externalStorage() const;
+
+    /**
+     * Sets the authentication configuration ID to be used for the current external storage (if
+     * defined)
+     * \since QGIS 3.22
+     */
+    void setStorageAuthConfigId( const QString &authCfg );
+
+    /**
+     * Returns the authentication configuration ID used for the current external storage (if defined)
+     * \since QGIS 3.22
+     */
+    const QString &storageAuthConfigId() const;
+
+    /**
+     * Set \a urlExpression expression, which once evaluated, provide the URL used to store selected
+     * documents. This is used only if an external storage has been defined
+     * \see setStorageType(), externalStorage()
+     * \since 3.22
+     */
+    void setStorageUrlExpression( const QString &urlExpression );
+
+    /**
+     * Returns the original, unmodified expression string, which once evaluated, provide the
+     * URL used to store selected documents. This is used only if an external storage has been defined.
+     * Returns null if no expression has been set.
+     * \see setStorageUrlExpression()
+     * \since 3.22
+     */
+    QString storageUrlExpressionString() const;
+
+    /**
+     * Returns expression, which once evaluated, provide the URL used to store selected
+     * documents. This is used only if an external storage has been defined.
+     * Returns null if no expression has been set.
+     * \see setStorageUrlExpression()
+     * \since 3.22
+     */
+    QgsExpression *storageUrlExpression() const;
```

Comment:
Why a pointer return here? The ownership is unclear

Reply:
It has no need for QGIS application, it's just in case, somebody need somehow to interact with the defined expression through the API. It mas mostly to be consistent wtih [what is done](https://github.com/qgis/QGIS/blob/master/src/core/qgsfeaturerequest.h#L415) in QgsFeatureRequest.

> The ownership is unclear

I would have said that when returning a pointer, the default expected behavior when nothing else is specified is that object keep ownership of the returned pointer, like it's done in [Qt](https://doc.qt.io/qt-5/qwidget.html#backingStore) for instance or [QGIS](https://github.com/qgis/QGIS/blob/master/src/core/qgsfeaturerequest.h#L415).

But, If you think it's necessary, I can add it.